libpod/runtime_pod_pause_linux.go

```diff
@@ -0,0 +1,234 @@
+)
+
+func (r *Runtime) buildPauseImage(ctx context.Context, p *Pod) (string, error) {
+	// Buildah new
+	// TODO make sure this matches global cli.context
+	signaturePolicy := ""
+	store := r.store
+	commonOpts := &buildah.CommonBuildOptions{}
+	transport := DefaultTransport
+	isolation := buildah.IsolationDefault
+	sc := &types.SystemContext{}
+	pullPolicy := buildah.PullNever
+	namespaceOptions := make(buildah.NamespaceOptions, 0, 7)
+
+	builderOptions := buildah.BuilderOptions{
+		FromImage:             "scratch",
+		Transport:             transport,
+		Container:             "pause",
+		PullPolicy:            pullPolicy,
+		SignaturePolicyPath:   signaturePolicy,
+		SystemContext:         sc,
+		DefaultMountsFilePath: "",
+		Isolation:             isolation,
+		NamespaceOptions:      namespaceOptions,
+		CommonBuildOpts:       commonOpts,
+	}
+
+	builderOptions.ReportWriter = os.Stderr
+
+	builder, err := buildah.NewBuilder(ctx, store, builderOptions)
+	if err != nil {
+		return "", err
+	}
+
+	// Buildah onBuild
+	addOptions := buildah.AddAndCopyOptions{
+		Chown:  "",
+		Hasher: digest.Canonical.Digester().Hash(),
+	}
+
+	err = builder.Add(pauseCmd, false, addOptions, pauseLocation)
+	if err != nil {
+		return "", err
+	}
+
+	builder.SetEntrypoint([]string{pauseCmd})
+	builder.SetCmd([]string{pauseCmd})
+
+	// Buildah commit
+	image := fmt.Sprintf("localhost/%s", pauseImgName)
+	compress := archive.Gzip
+	timestamp := time.Now().UTC()
+	format := buildah.OCIv1ImageManifest
+
+	dest, err := alltransports.ParseImageName(image)
+	if err != nil {
+		candidates := util.ResolveName(image, "", sc, store)
+		if len(candidates) == 0 {
+			return "", errors.Errorf("error parsing target image name %q", image)
+		}
+		dest2, err2 := storage.Transport.ParseStoreReference(store, candidates[0])
+		if err2 != nil {
+			return "", errors.Wrapf(err, "error parsing target image name %q", image)
+		}
+		dest = dest2
+	}
+
+	commitOptions := buildah.CommitOptions{
+		PreferredManifestType: format,
+		Compression:           compress,
+		SignaturePolicyPath:   "",
+		HistoryTimestamp:      &timestamp,
+		SystemContext:         sc,
+		Squash:                false,
+	}
+
+	return builder.Commit(ctx, dest, commitOptions)
+}
+
+func (r *Runtime) getPauseImage(ctx context.Context, p *Pod) (*image.Image, error) {
+	// First, check if the pause image is already in storage. If so, use it.
+	newImage, err := r.ImageRuntime().New(ctx, pauseImgName, "", "", nil, nil, image.SigningOptions{}, false, false)
+	if err != nil {
+		// If not, generate pause image
+		imgID, err2 := r.buildPauseImage(ctx, p)
+		if err2 != nil {
+			return nil, err
+		}
+		newImage, err = r.ImageRuntime().New(ctx, imgID, "", "", os.Stderr, nil, image.SigningOptions{}, false, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newImage, nil
+}
+
+func (r *Runtime) getPauseContainer(ctx context.Context, p *Pod, imgName, imgID string) (*Container, error) {
```

Comment:
This shouldn't do so much on its own - reuse the existing runtime.NewContainer()

Reply:
We should generate a spec, throw it into NewContainer(), and use the container that is returned as the infra container

Reply:
I basically copied the NewContainer code but simplified it. The reason  I didn't call NewContainer() directly is because it is a locked function and we already have the lock. And newContainer already exists.

Reply:
nevermind, newContainer isn't scoped to runtime. I will fix that.

Reply:
`newContainer` should probably be refactored into `runtime.NewContainer()` - it doesn't really do much, and is only called from `runtime.NewContainer()`

Reply:
ok I'll add newContainer to runtime.newContainer, which will be called in runtime.NewContainer